src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/Dataflow/MethodBodyScanner.cs

```diff
@@ -942,5 +988,87 @@ protected virtual void HandleStoreParameter(MethodIL method, int offset, int ind
             int offset,
             ValueNodeList methodParams,
             out ValueNode methodReturnValue);
+
+        // Limit tracking array values to 32 values for performance reasons. There are many arrays much longer than 32 elements in .NET, but the interesting ones for the linker are nearly always less than 32 elements.
+        private const int MaxTrackedArrayValues = 32;
```

Comment:
This reminds me of .NET Native data flow analyzer that was full of arbitrary cut-offs like this...

Reply:
I asked for an arbitrary limit because the other option is that we would model the values of arbitrarily big arrays and risk running out of memory while doing so.

https://github.com/mono/linker/pull/1955#discussion_r611645628

Maybe it would have been better not to model arrays at all...

Reply:
> Maybe it would have been better not to model arrays at all...

Yep. We are getting on a slippery slope with this.

Reply:
> > Maybe it would have been better not to model arrays at all...
> 
> Yep. We are getting on a slippery slope with this.

Cc @vitek-karas

Reply:
Alternative would be to only model arrays we care about. Which is currently almost exclusively `Type[]`. We could increase the limit (a lot) because nobody will create large arrays of Type statically. My understanding is that the worry with the existing behavior is that it catches on things like `char[]` which can be very large but still fully static.

Reply:
> nobody will create large arrays of Type statically.

Famous last words

Reply:
OK OK - I knew this is going to come back to bite me... 😄 
But realistically - we only need to track arrays which are `Type[]` and their length is as long as the largest number of generic parameters on a single type/method. I don't know if reflection/IL has some limits on that, probably not really, so we might as well not put a limit on it either in the linker. If somebody does that ... and linker runs out of memory somehow... I mean there are probably easier ways to force linker to consume lot of memory.